// ls-user/user.js
/*
*  JavaScript User object implementation for LibreSignage.
*  This class is used to load and manipulate user data.
*/

var Quota = require('./quota.js').Quota;

var APIEndpoints = require('libresignage/api/APIEndpoints');
var Assert = require('libresignage/util/assert/Assert');

class User {
	constructor(api) {
		Assert.assert(api != null, 'An API object must be supplied.');
		this.api = api;
		this.data = null;
	}

	async create(user) {
		this.data = {};
		let resp = await this.api.call(
			APIEndpoints.USER_CREATE,
			{'user': user}
		);

		this.data = Object.assign(this.data, resp.user);
		this.data.pass = resp.pass;
	}

	async load(user) {
		let resp = null;
		this.data = {};

		if (user == null) {
			resp = await this.api.call(
				APIEndpoints.USER_GET_CURRENT,
				null
			);
		} else {
			resp = await this.api.call(
				APIEndpoints.USER_GET,
				{'user': user}
			);
		}

		/*
		*  Create a separate Quota object from the quota data
		*  if the data exists in the response. This check is
		*  needed because quota data is only returned if the
		*  user is requesting their own userdata.
		*/
		if ('quota' in resp.user) {
			this.data = { quota: new Quota(resp.user.quota) };
			delete resp.user.quota;
		}

		// Add rest of the fetched data as-is.
		this.data = Object.assign(this.data, resp.user);
	}

	async save() {
		Assert.assert(this.data != null, "User not loaded.");
		await this.api.call(
			APIEndpoints.USER_SAVE,
			{
				'user':   this.data.user,
				'groups': this.data.groups,
				'pass':   this.data.pass ? this.data.pass : null
			}
		);
	}

	async remove() {
		Assert.assert(this.data != null, "User not loaded.");
		await this.api.call(
			APIEndpoints.USER_REMOVE,
			{user: this.data.user}
		);
		this.data = null;
	}

	set_password(pass) { this.data.pass = pass;     }
	set_groups(groups) { this.data.groups = groups; }

	get_user()         { return this.data.user;     }
	get_password()     { return this.data.pass;     }
	get_groups()       { return this.data.groups;   }
	get_sessions()     { return this.data.sessions; }
	get_quota()        { return this.data.quota;    }

	/**
	* Check whether a user is in a group. If 'group' is an array,
	* this function checks whether a user is in any of the groups
	* in 'group'.
	*
	* @param {string|string[]} group The group(s) to check.
	*
	* @return True if the user is in the group(s) and false otherwise.
	*/
	is_in_group(group) {
		if (Array.isArray(group)) {
			for (let g in group) {
				if (this.data.groups.includes(g)) { return true; }
			}
			return false;
		} else {
			return this.data.groups.includes(group);
		}
	}

	/**
	* Get an array of all existing usernames via the API.
	*
	* @param {APIInterface} api An APIInterface object.
	*
	* @return {string[]} An array of usernames.
	*/
	static async list_all(api) {
		Assert.assert(api != null, 'An API object must be supplied.');
		let resp = await api.call(
			APIEndpoints.USERS_LIST,
			null
		);
		return resp.users;
	}

	/**
	* Get all existing User objects via the API.
	*
	* @param {APIInterface} api An APIInterface object.
	*
	* @return {User[]} An array of user objects.
	*/
	static async get_all(api) {
		Assert.assert(api != null, 'An API object must be supplied.');
		let users = {};
		let names = await User.list_all(api);
		for (let n of names ) {
			users[n] = new User(api);
			await users[n].load(n);
		}
		return users;
	}
}
exports.User = User;
