// ls-markup/error.js
/*
*  Error definitions for the LibreSignage markup parser.
*/

module.exports.ECODES = {
	'EEOL': {
		code: 0,
		msg: 'Unexpected End Of File'
	},
	'ETOK': {
		code: 1,
		msg: 'Unexpected token'
	},
	'ESYN': {
		code: 2,
		msg: 'Invalid syntax'
	},
	'EEXP': {
		code: 3,
		msg: 'Invalid expression'
	},
	'ETAG': {
		code: 4,
		msg: 'Invalid tag'
	},
	'ETYPE': {
		code: 5,
		msg: 'Type error'
	},
	'EARG': {
		code: 6,
		msg: 'Argument error'
	},
	'EVAL': {
		code: 7,
		msg: 'Value error'
	}
};

module.exports.MarkupSyntaxError = class MarkupSyntaxError extends Error {
	constructor(code, lexeme, message, file, line) {
		super(message, file, line);
		if (Error.captureStackTrace) {
			Error.captureStackTrace(this, MarkupSyntaxError);
		}

		let ecodes = module.exports.ECODES;
		this.code = ecodes[code];
		this.message = message;
		this.lexeme = lexeme;
	}

	line() { return this.lexeme.ln_num; }
	char() { return this.lexeme.ln_pos; }

	toString(ln_base) {
		let base = (!ln_base) ? 0: ln_base;
		let ret = `${this.code.msg}: ${this.message}`;

		if (this.lexeme) {
			ret += ` (@ ${base + this.lexeme.ln_num}:` +
					`${this.lexeme.ln_pos})`;
		}
		return ret;
	}
}
