// ls-markup/eval_defs.js
/*
*  LibreSignage markup syntax definitions.
*/

/*
*  Evaluator argument type defs.
*/
const ARGTYPES = {
	'percent': {
		code: 0,
		typestr: 'percent'
	},
	'int': {
		code: 1,
		typestr: 'int'
	},
	'str': {
		code: 2,
		typestr: 'str'
	},
	'rstr': {
		code: 3,
		typestr: 'rstr'
	}
};
module.exports.ARGTYPES = ARGTYPES;

/*
*  Evaluator expression definitions.
*
*  Each entry in this object defines a valid tag of the
*  form [name][/name] where 'name' is the key in the object.
*  Each definition contain the following values:
*    - has:    A dictionary of required arguments. The arguments
*              are defined as key-value pairs where the key is the
*              argument name and the value is one of the argument
*              types defined in the ARGTYPES dictionary.
*    - handle: A function for generating the output corresponding
*              to the expression. The arguments parsed from the
*              transpiler input are passed to this function as the
*              first argument (a dictionary). This function must
*              return an array containing two strings. The first
*              one is the opening tag (eg. [name arg=1...]) and
*              the second one is the closing tag (eg. [/name]).
*/
module.exports.EXPRESSIONS = {
	root: { has: {}, handle: () => { return ['', '']; } },
	rstring: {
		has: {'text': ARGTYPES.rstr},
		handle: (arg) => { return [arg.text, '']; }
	},
	h: {
		has: { 'size': ARGTYPES.percent },
		handle: (arg) => {
			let s = arg.size.replace('%', 'vh');
			return [
				`<h1 style="font-size: ${s};">`,
				'</h1>'
			];
		}
	},
	lead: {
		has: {},
		handle: (arg) => {
			return [
				`<p class="lead">`,
				'</p>'
			]
		}
	},
	b: {
		has: {},
		handle: () => {
			return [
				'<span style="font-weight: bold;">',
				'</span>'
			];
		}
	},
	i: {
		has: {},
		handle: () => {
			return [
				'<span style="font-style: italic;">',
				'</span>'
			];
		}
	},
	img: {
		has: {
			'url': ARGTYPES.str,
			'width': ARGTYPES.percent,
			'height': ARGTYPES.percent
		},
		handle: (arg) => {
			let w = arg.width.replace('%', 'vh');
			let h = arg.height.replace('%', 'vw');
			return [
				`<img src=${arg.url} style="width: ${w}; height: ${h};">`,
				'</img>'
			];
		}
	},
	p: {
		has: {},
		handle: () => {
			return ['<p>', '</p>'];
		}
	},
	size: {
		has: { 'size': ARGTYPES.percent },
		handle: (arg) => {
			let s = arg.size.replace('%', 'vh');
			return [
				`<span style="font-size: ${s}">`,
				'</span>'
			];
		}
	},
	color: {
		has: { 'c': ARGTYPES.rstr },
		handle: (arg) => {
			return [
				`<span style="color: ${arg.c};">`,
				'</span>'
			];
		}
	},
	container: {
		has: {
			'top': ARGTYPES.percent,
			'right': ARGTYPES.percent,
			'bottom': ARGTYPES.percent,
			'left': ARGTYPES.percent
		},
		handle: (arg) => {
			let t = arg.top.replace('%', 'vh');
			let r = arg.right.replace('%', 'vh');
			let b = arg.bottom.replace('%', 'vh');
			let l = arg.left.replace('%', 'vh');
			return [
				`<div style="padding: ${t} ${r} ${b} ${l};">`,
				'</div>'
			];
		}
	},
	xcenter: {
		has: {},
		handle: () => {
			return [
				`<div style="margin-left: auto;
							margin-right: auto;
							text-align: center;
							width: auto;
							height: auto;">`,
				'</div>'
			];
		}
	},
	columns: {
		has: {},
		handle: () => {
			return [
				`<div style="display: flex;
						flex-direction: row;
						width: auto;
						height: auto;">`,
				'</div>'
			];
		}
	},
	align: {
		has: { 'type': ARGTYPES.rstr },
		accept: { 'type': ['left', 'right', 'center', 'justify']},
		handle: (arg) => {
			return [
				`<div style="text-align: ${arg.type};">`,
				'</div>'
			];
		}
	}
};
