// ls-multiselect/multiselect.js
/*
*  Control functionality for the multiselect element.
*/

var $ = require('jquery');
var val = require('ls-validator');

const MS_TEMPLATE = (id, ) => `
	<div id="${id}" class="multiselect container-fluid">
		<div class="ms-controls container-fluid">
			<input class="ms-input col form-control" type="text">
			<button class="ms-add col btn btn-primary" type="button">
				<i class="fas fa-plus-circle"></i>
			</button>
			<div class="invalid-feedback"></div>
		</div>
		<div class="ms-values container-fluid"></div>
	</div>
`;

exports.MultiSelect = class MultiSelect {
	constructor(dest_id, id, validators, nodups) {
		if (!dest_id) {
			throw new Error('No destination id.');
		}
		if (!id) {
			throw new Error('Invalid empty ID for multiselector.');
		}

		// Create the HTML elements and add them to the document.
		$(`#${dest_id}`).append(MS_TEMPLATE(id));

		this.enabled = true;
		this.val_valid = true;
		this.selected = [];

		this.nodups = nodups;
		this.root = $(`#${id}`);
		this.input = $(`#${id} > .ms-controls > .ms-input`);
		this.btn_add = $(`#${id} > .ms-controls > .ms-add`);
		this.values = $(`#${id} > .ms-values`);

		// Add a listener for Enter keypresses.
		this.input.on('keypress', (event) => {
			if (event.key == 'Enter' && this.val_valid) {
				this.add(this.input.val());
				this.input.val('');
			}
		});

		// Add a listener for the (+) button.
		this.btn_add.on('click', () => {
			this.add(this.input.val());
			this.input.val('');
		});

		// Add validators for the input.
		if (validators && validators.length) {
			this.vtrig = new val.ValidatorTrigger(
				[new val.ValidatorSelector(
					this.input,
					this.root,
					validators,
					null
				)],
				(valid) => {
					this.val_valid = valid;
					this.btn_add.prop(
						'disabled',
						!valid
					);
				}
			);
		}
	}

	add(option) {
		/*
		*  Select a new option.
		*/
		if (this.nodups && this.selected.includes(option)) {
			return;
		}
		if (!option) { return; }

		var cont = $('<div>', {
			'id': `ms-opt-${option}`,
			'class': 'ms-val',
			'text': option
		});
		var rm = $('<span>', {
			'class': 'ms-rm fas fa-times'
		});
		cont.append(rm);
		rm.on('click', () => { this.remove(option); });

		this.selected.push(option);
		this.values.append(cont);

	}

	set(options) {
		/*
		*  Select all the options in 'options'.
		*/

		// Clear existing selections.
		this.selected = [];
		this.values.html('');

		// Add new selections.
		for (let o of options) { this.add(o); }
	}

	remove(option) {
		/*
		*  Remove the selection 'option'.
		*/
		if (!this.selected.includes(option)) {
			throw new Error('Option not selected.');
		}
		$(`#ms-opt-${option}`).remove();
		this.selected.splice(this.selected.indexOf(option), 1);
	}

	remove_all() {
		for (let o in Object.keys(this.selected)) {
			this.remove(o);
		}
	}

	enable() {
		/*
		*  Enable the MultiSelect.
		*/
		this.enabled = true;
		this.values.css('background-color', 'white');
		this.input.prop('disabled', false);
		this.root.find('.ms-rm').css('display', 'inline');

		// This also enables the add button if needed.
		for (let s of this.vtrig.selectors) { s.enable(); }
	}

	disable() {
		/*
		*  Disable the MultiSelect.
		*/
		this.enabled = false;
		for (let s of this.vtrig.selectors) { s.disable(); }
		this.values.css('background-color', 'var(--gray-3)');
		this.input.prop('disabled', true);
		this.btn_add.prop('disabled', true);
		this.root.find('.ms-rm').css('display', 'none');
	}
}
