// ls-queue/queue.js
/*
*  Queue class definition. This class is used to manipulate
*  LibreSignage Slide queus. The Queue class uses SlideLists
*  underneath.
*/

var Slide = require('ls-slide').Slide;
var SlideList = require('ls-slidelist').SlideList;

var APIEndpoints = require('libresignage/api/APIEndpoints');
var Assert = require('libresignage/util/assert/Assert');

class Queue {
	constructor(api) {
		Assert.assert(api != null, "API object can't be null.");
		this.api = api;
		this.data = null;
	}

	async load(name) {
		/*
		*  Load the Queue 'name' along with all of its Slides.
		*/
		let ld = [];
		let resp = await this.api.call(
			APIEndpoints.QUEUE_GET,
			{'name': name}
		);
		this.data = {
			'name': name,
			'owner': resp.queue.owner,
			'slides': new SlideList()
		};

		for (let s of resp.queue.slide_ids) {
			this.data.slides.get_slides()[s] = new Slide(this.api);
			ld.push(this.data.slides.get_slides()[s].load(s, false, false));
		}
		return Promise.all(ld);
	}

	async create(name) {
		/*
		*  Create a new empty queue.
		*/
		await this.api.call(
			APIEndpoints.QUEUE_CREATE,
			{ 'name': name }
		);
		await this.load(name);
	}

	async remove() {
		Assert.assert(
			this.data != null
			&& this.data.name != null,
			"No queue loaded."
		);
		await this.api.call(
			APIEndpoints.QUEUE_REMOVE,
			{ 'name': this.data.name }
		);
		this.data = null;
	}

	async update() { await this.load(this.data.name); }

	has_slide(id) {
		for (let s of Object.values(this.data.slides.get_slides())) {
			if (s.get('id') === id) { return true; }
		}
		return false;
	}

	get_name()   { return this.data.name;   }
	get_owner()  { return this.data.owner;  }
	get_slides() { return this.data.slides; }

	static async get_queues(api) {
		return (await api.call(APIEndpoints.QUEUE_LIST, null)).queues;
	}
}
exports.Queue = Queue;
