// ls-queue/queue.js
/*
*  Queue class definition. This class is used to manipulate
*  LibreSignage Slide queus. The Queue class uses SlideLists
*  underneath.
*/

var Slide = require('ls-slide').Slide;
var SlideList = require('ls-slidelist').SlideList;
var assert = require('ls-assert').assert;
var APIEndpoints = require('ls-api').APIEndpoints;

class Queue {
	constructor(api) {
		assert(api != null, "API object can't be null.");
		this.api = api;
		this.data = null;
	}

	async load(name) {
		/*
		*  Load the Queue 'name' along with all of its Slides.
		*/
		let ld = [];
		let resp = await this.api.call(
			APIEndpoints.QUEUE_GET,
			{'name': name}
		);
		this.data = {
			'name': name,
			'owner': resp.owner,
			'slides': new SlideList()
		};

		for (let s of Object.values(resp.slides)) {
			this.data.slides.get_slides()[s.id] = new Slide(this.api);
			ld.push(this.data.slides.get_slides()[s.id].load(
				s.id, false, false
			));
		}
		return Promise.all(ld);
	}

	async update() { await this.load(this.name); }

	get_name()   { return this.data.name;   }
	get_owner()  { return this.data.owner;  }
	get_slides() { return this.data.slides; }

	static async get_queues(api) {
		return (await api.call(APIEndpoints.QUEUE_LIST, null)).queues;
	}
}
exports.Queue = Queue;
