// libresignage/queue/Queue/Queue.js
/**
* @module libresignage/queue/Queue
*/

var SlideList = require('ls-slidelist').SlideList;

var APIEndpoints = require('libresignage/api/APIEndpoints');
var Assert = require('libresignage/util/assert/Assert');
var Slide = require('libresignage/slide/Slide');

/**
* Class representing a slide queue.
*/
class Queue {
	/**
	* Construct a Queue object.
	*
	* @param {string} api An initialized APIInterface object.
	*
	* @throws {AssertError} If api == null.
	*/
	constructor(api) {
		Assert.assert(api != null, "API object can't be null.");
		this.api = api;
		this.data = null;
	}

	/**
	* Load a queue and all of its slides via the API.
	*
	* @param {string} name The name of the queue to load.
	*
	* @return {Promise} A Promise that resolves once all the slides are loaded.
	*/
	async load(name) {
		let ld = [];
		let resp = await this.api.call(
			APIEndpoints.QUEUE_GET,
			{'name': name}
		);
		this.data = {
			'name': name,
			'owner': resp.queue.owner,
			'slides': new SlideList()
		};

		for (let s of resp.queue.slide_ids) {
			this.data.slides.get_slides()[s] = new Slide(this.api);
			ld.push(this.data.slides.get_slides()[s].load(s, false, false));
		}
		return Promise.all(ld);
	}

	/**
	* Create a new empty queue via the API.
	*
	* @param {string} name The name of the new queue.
	*/
	async create(name) {
		await this.api.call(
			APIEndpoints.QUEUE_CREATE,
			{ 'name': name }
		);
		await this.load(name);
	}

	/**
	* Remove the loaded queue via the API.
	*
	* @throws {AssertError} If no queue is loaded.
	*/
	async remove() {
		Assert.assert(
			this.data != null
			&& this.data.name != null,
			"No queue loaded."
		);
		await this.api.call(
			APIEndpoints.QUEUE_REMOVE,
			{ 'name': this.data.name }
		);
		this.data = null;
	}

	/**
	* Refresh the loaded queue data via the API.
	*/
	async update() { await this.load(this.data.name); }

	/**
	* Check whether the loaded queue has a slide.
	*
	* @param {string} id The slide ID to search for.
	*
	* @return {boolean} True if the matching slide exists, false otherwise.
	*/
	has_slide(id) {
		for (let s of Object.values(this.data.slides.get_slides())) {
			if (s.get('id') === id) { return true; }
		}
		return false;
	}

	/**
	* Get the name of a queue.
	*
	* @return {string|null} The name of the queue.
	*/
	get_name()   { return this.data.name;   }

	/**
	* Get the owner of a queue.
	*
	* @param {string} The owner of the queue.
	*/
	get_owner()  { return this.data.owner;  }

	/**
	* Get the internal SlideList object of a queue.
	*
	* @return {SlideList} The SlideList object.
	*/
	get_slides() { return this.data.slides; }

	/**
	* Get a list of all queue names via the API.
	*
	* @return {string[]} An array of queue names.
	*/
	static async get_queues(api) {
		return (await api.call(APIEndpoints.QUEUE_LIST, null)).queues;
	}
}
module.exports = Queue;
