// libresignage/queue/Queue/Queue.js
var SlideArray = require('libresignage/queue/SlideArray');
var APIEndpoints = require('libresignage/api/APIEndpoints');
var Assert = require('assert');
var Slide = require('libresignage/slide/Slide');

/**
* Class representing a slide queue.
*
* @memberof module:libresignage/queue
*/
class Queue {
	/**
	* Construct a Queue object.
	*
	* @param {string} api An initialized APIInterface object.
	*
	* @throws {AssertionError} If api == null.
	*/
	constructor(api) {
		Assert.ok(api != null, "API object can't be null.");
		this.api = api;

		this.name = null;
		this.owner = null;
		this.slidearray = null;
	}

	/**
	* Load a queue and all of its slides via the API.
	*
	* @param {string} name The name of the queue to load.
	*
	* @return {Promise} A Promise that resolves once all the slides are loaded.
	*
	* @throws {AssertionError} If name == null.
	*/
	async load(name) {
		Assert.ok(name != null, "'name' cannot be null.");
		let promises = [];
		let resp = await this.api.call(
			APIEndpoints.QUEUE_GET,
			{'name': name}
		);

		this.name = name;
		this.owner = resp.queue.owner;
		this.slidearray = new SlideArray();

		for (let s of resp.queue.slide_ids) {
			let tmp = new Slide(this.api);
			promises.push(tmp.load(s, false, false));
			this.slidearray.push(tmp);
		}

		return Promise.all(promises);
	}

	/**
	* Create a new empty Queue via the API.
	*
	* @param {string} name The name of the new !ueue.
	*
	* @throws {AssertionError} If 'name' == null.
	*/
	async create(name) {
		Assert.ok(name != null, "'name' cannot be null.");
		await this.api.call(
			APIEndpoints.QUEUE_CREATE,
			{'name': name}
		);
		await this.load(name);
	}

	/**
	* Remove a Queue via the API.
	*
	* @throws {AssertionError} If no Queue is loaded.
	*/
	async remove() {
		Assert.ok(this.name != null, "No Queue loaded.");
		await this.api.call(
			APIEndpoints.QUEUE_REMOVE,
			{'name': this.name}
		);
	}

	/**
	* Refresh a !ueue data via the API.
	*
	* @throws {AssertionError} If no Queue is loaded.
	*/
	async update() {
		Assert.ok(this.name != null, "No Queue loaded.");
		await this.load(this.name);
	}

	/**
	* Check whether a Queue contains a Slide.
	*
	* @param {string} id The Slide ID to search for.
	*
	* @return {boolean} True if the matching Slide exists, false otherwise.
	*
	* @throws {AssertionError} If no Queue is loaded.
	*/
	has_slide(id) {
		Assert.ok(this.name != null, "No Queue loaded.");
		for (let s of this.slidearray) {
			if (s.get('id') === id) {
				return true;
			}
		}
		return false;
	}

	/**
	* Get the name of a Queue.
	*
	* @return {string|null} The name of the Queue.
	*/
	get_name() { return this.name; }

	/**
	* Get the owner of a Queue.
	*
	* @param {string|null} The owner of the Queue.
	*/
	get_owner() { return this.owner; }

	/**
	* Get the SlideArray object of a Queue.
	*
	* @return {SlideList|null} The SlideArray object.
	*/
	get_slidearray() { return this.slidearray; }

	/**
	* Get a list of all Queue names via the API.
	*
	* @return {string[]} An array of Queue names.
	*/
	static async get_queues(api) {
		return (await api.call(APIEndpoints.QUEUE_LIST, {})).queues;
	}
}
module.exports = Queue;
