// ls-markup-new/tokens.js
var lexeme = require('./lexeme.js');
var mutil = require('./mutil.js');

class TokenMatcher {
	constructor(category, type, regex, matcher) {
		this.category = category;
		this.type = type;
		this.regex = regex;
		if (!matcher) {
			this.matcher = this.def_matcher;
		} else {
			this.matcher = matcher;
		}
	}

	def_matcher(str, at) {
		let m = str.substring(at).match(this.regex);
		if (m) {
			return new lexeme.Lexeme(
				this.category,
				this.type,
				m[0]
			);
		} else {
			return null;
		}
	}
}

module.exports.TOKENS = [
	new TokenMatcher('separator', 'BRACK_OPEN', /^\[/),
	new TokenMatcher('separator', 'BRACK_CLOSE', /^\]/),
	new TokenMatcher(
		'separator', 'BRACK_SLASH', /^\//,
		function (str, at, index, lexemes) {
			let tmp = {};
			tmp[index - 1] = new lexeme.Lexeme('separator', 'BRACK_OPEN');
			if (!mutil.chk_lexeme_context(lexemes, tmp)) {
				return null;
			} else {
				return this.def_matcher(str, at);
			}
		}
	),
	new TokenMatcher('separator', 'WHITESPACE', /^\s+/),
	new TokenMatcher(
		'operator', 'OP_ASSIGN', /^=/,
		function(str, at, index, lexemes) {
			let tmp = {};
			let ret = null;
			tmp[index - 1] = new lexeme.Lexeme('literal', 'LITERAL');
			if (!mutil.chk_lexeme_context(lexemes, tmp)) {
				return null;
			} else {
				let ret = this.def_matcher(str, at);
				if (ret) {
					// Convert the LITERAL into VARNAME.
					lexemes[index - 1] = new lexeme.Lexeme(
						'identifier',
						'VARNAME',
						lexemes[index - 1].raw
					);
				}
				return ret;
			}
		}
	),
	new TokenMatcher(
		'identifier', 'TAGNAME', /^[A-Za-z0-9_-]+/,
		function(str, at, index, lexemes) {
			let tmp_1 = {};
			let tmp_2 = {};

			tmp_1[index - 1] = new lexeme.Lexeme(
				'separator', 'BRACK_OPEN'
			);

			tmp_2[index - 2] = new lexeme.Lexeme(
				'separator', 'BRACK_OPEN'
			);
			tmp_2[index - 1] = new lexeme.Lexeme(
				'separator', 'BRACK_SLASH'
			);

			if (
				!mutil.chk_lexeme_context(lexemes, tmp_1)
				&& !mutil.chk_lexeme_context(lexemes, tmp_2)
			) {
				return null;
			} else {
				return this.def_matcher(str, at);
			}
		}
	),
	new TokenMatcher(
		'identifier', 'VARNAME', null,
		function() { return null; }
	),
	new TokenMatcher('literal', 'LITERAL', /^[^=\[\]\s]+/)
];

module.exports.get_tok = function(category, type) {
	for (let tm of module.exports.TOKENS) {
		if (
			tm.category == category
			&& tm.type == type
		) { return tm; }
	}
	return null;
}
