// ls-api/endpoints.js
const API_BASE = `/api/endpoint`;

exports.APIMime = {
	JSON: 'application/json',
	FORM_DATA: 'multipart/form-data',
	X_WWW_FORM_URLENCODED: 'application/x-www-form-urlencoded'
}

exports.APIEndpoints = {
	// -- User management API endpoints --
	USER_GET_QUOTA: {
		uri:      `${API_BASE}/user/user_get_quota.php`,
		method:   "GET",
		auth:     true
	},
	USER_CREATE: {
		uri:      `${API_BASE}/user/user_create.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	USER_REMOVE: {
		uri:      `${API_BASE}/user/user_remove.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	USER_SAVE: {
		uri:      `${API_BASE}/user/user_save.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	USER_GET: {
		uri:      `${API_BASE}/user/user_get.php`,
		method:   "GET",
		auth:     true
	},
	USER_GET_CURRENT: {
		uri:      `${API_BASE}/user/user_get_current.php`,
		method:   "GET",
		auth:     true
	},
	USERS_GET_ALL: {
		uri:      `${API_BASE}/user/users_get_all.php`,
		method:   "GET",
		auth:     true
	},
	USERS_LIST: {
		uri:      `${API_BASE}/user/users_list.php`,
		method:   "GET",
		auth:     true
	},

	// -- Slide API endpoints --
	SLIDE_LIST: {
		uri:      `${API_BASE}/slide/slide_list.php`,
		method:   "GET",
		auth:     true
	},
	SLIDE_GET: {
		uri:      `${API_BASE}/slide/slide_get.php`,
		method:   "GET",
		auth:     true,
	},
	SLIDE_SAVE: {
		uri:      `${API_BASE}/slide/slide_save.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	SLIDE_DUP: {
		uri:      `${API_BASE}/slide/slide_dup.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	SLIDE_RM: {
		uri:      `${API_BASE}/slide/slide_rm.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	SLIDE_LOCK_ACQUIRE: {
		uri:      `${API_BASE}/slide/slide_lock_acquire.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	SLIDE_LOCK_RELEASE: {
		uri:      `${API_BASE}/slide/slide_lock_release.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	SLIDE_UPLOAD_ASSET: {
		uri:      `${API_BASE}/slide/asset/slide_upload_asset.php`,
		method:   "POST",
		req_mime: "multipart/form-data",
		auth:     true
	},			
	SLIDE_REMOVE_ASSET: {
		uri:      `${API_BASE}/slide/asset/slide_remove_asset.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},			
	SLIDE_GET_ASSET: {
		uri:      `${API_BASE}/slide/asset/slide_get_asset.php`,
		method:   "GET",
		auth:     true
	},
	SLIDE_GET_ASSET_THUMB: {
		uri:      `${API_BASE}/slide/asset/slide_get_asset_thumb.php`,
		method:   "GET",
		auth:     true
	},

	// -- Queue API endpoints --
	QUEUE_GET: {
		uri:      `${API_BASE}/queue/queue_get.php`,
		method:   "GET",
		auth:     true
	},
	QUEUE_CREATE: {
		uri:      `${API_BASE}/queue/queue_create.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	QUEUE_LIST: {
		uri:      `${API_BASE}/queue/queue_list.php`,
		method:   "GET",
		auth:     true
	},
	QUEUE_REMOVE: {
		uri:      `${API_BASE}/queue/queue_remove.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},

	// -- Authentication API endpoints --
	AUTH_LOGIN: {
		uri:      `${API_BASE}/auth/auth_login.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     false
	},
	AUTH_LOGOUT: {
		uri:      `${API_BASE}/auth/auth_logout.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	AUTH_LOGOUT_OTHER: {
		uri:      `${API_BASE}/auth/auth_logout_other.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	AUTH_SESSION_RENEW: {
		uri:      `${API_BASE}/auth/auth_session_renew.php`,
		method:   "POST",
		req_mime: "application/json",
		auth:     true
	},
	AUTH_GET_SESSIONS: {
		uri:       `${API_BASE}/auth/auth_get_sessions.php`,
		method:   "GET",
		auth:     true
	},
	AUTH_GET_DATA: {
		uri:      `${API_BASE}/auth/auth_get_data.php`,
		method:   "GET",
		auth:     true
	},

	// -- General information API endpoints --
	API_ERR_CODES: {
		uri:      `${API_BASE}/general/api_err_codes.php`,
		method:   "GET",
		auth:     false
	},
	API_ERR_MSGS: {
		uri:      `${API_BASE}/general/api_err_msgs.php`,
		method:   "GET",
		auth:     false
	},
	SERVER_LIMITS: {
		uri:      `${API_BASE}/general/server_limits.php`,
		method:   "GET",
		auth:     false
	},
	VER_INFO: {
		uri:      `${API_BASE}/general/ver_info.php`,
		method:   "GET",
		auth:     false
	}
};
