// ls-api/error.js
var assert = require('ls-assert');
var APIEndpoints = require('./endpoints').APIEndpoints;

class APIError extends Error{
	constructor(code, details, file, line) {
		super(details, file, line);
		if (Error.captureStackTrace) {
			Error.captureStackTrace(this, APIError);
		}
		this.code = code;
		this.details = details;
		if (this.code in APIError.messages) {
			this.message = APIError.messages[this.code].short + ' - ' +
						APIError.messages[this.code].long +
						(details ? ' (' + details + ')' : '');
		} else {
			this.message = 'Unknown API error. ' +
						(details ? ' (' + details + ')' : '');
		}
	}

	static err(name) {
		/*
		*  Get the integer error code for the error 'name'. This function
		*  is not normally needed, but it should be used in code where
		*  the error codes may not have been loaded yet. If the error
		*  'name' is not found, -1 is returned.
		*/
		return name in APIError.codes ? APIError.codes[name] : -1
	}

	static async init(api) {
		/*
		*  Initialize the API error system.
		*/
		if (!APIError.inited) {
			let tmp = {};
			tmp = await api.call(APIEndpoints.API_ERR_CODES, null);
			APIError.codes = tmp.codes;

			tmp = await api.call(APIEndpoints.API_ERR_MSGS, null);
			APIError.messages = tmp.messages;

			APIError.inited = true;
		}
	}
}
APIError.codes    = [];
APIError.messages = [];
APIError.inited   = false;
exports.APIError  = APIError;
