// libresignage/ui/components/DropSelect/DropSelect.js
var Assert = require('libresignage/util/assert/Assert');

/**
* A UI component that allows the user to make a selection from a list.
*
* @memberof module:libresignage/ui/components
*/
class DropSelect {
	/**
	* Construct a new DropSelect object.
	*
	* The container provided to this function *must* have an ID.
	*
	* @param {HTMLElement} container The container where the DropSelect
	*                                is created.
	*/
	constructor(container) {
		Assert.assert(container.id != '', 'Container element must have an ID.');

		this.container = container;
		this.options   = null;
		this.selected  = null;

		this.container.classList.add('dropdown', 'dropselect')
		this.container.innerHTML = DropSelect.make_html(this.container.id);

		// Attach open/close event listeners.
		this.container.addEventListener(('show.bs.dropdown'), e => {
			this.container.dispatchEvent('component.dropselect.show')
		});
		this.container.addEventListener(('hide.bs.dropdown'), e => {
			this.container.dispatchEvent('component.dropselect.hide')
		});
	}

	/**
	* Create the DropSelect HTML from a template.
	*
	* @param {string} container_id The ID of the container where the DropSelect
	*                              is created.
	*
	* @returns {string} The generated HTML.
	*/
	static make_html(container_id) {
		return `
			<button id="btn-${container_id}-dropselect"
				type="button"
				class="btn dropdown-toggle dropselect-open"
				data-toggle="dropdown"
				aria-haspopup="true"
				aria-expanded="false">
			</button>
			<div class="dropdown-menu dropselect-menu"
				aria-labelledby="btn-${container_id}-dropselect">
			</div>
		`;
	}

	/**
	* Set the options of a DropSelect.
	*
	* @param {string[]}    options An array of options.
	* @param {string|null} def     The default option to select.
	*/
	set_options(options, def) {
		this.options  = options;
		this.selected = null;

		let dropselect_menu = this.container.querySelector('.dropselect-menu');
		dropselect_menu.innerHTML = '';

		// Show the '< Empty >' item for empty lists.
		if (options.length === 0) {
			dropselect_menu.innerHTML += `
				<button class="dropdown-item"
						type="button"
						disabled>
					< Empty >
				</button>
			`;
			return;
		}

		// Add selector options from 'options'.
		for (let o of options) {
			let template = document.createElement('template');
			template.innerHTML = `
				<button class="dropdown-item"
						type="button"
						value="${o}">
					${o}
				</button>
			`;
			template.content.addEventListener('click', () => {
				this.select(o, true)
			});
			dropselect_menu.innerHTML += template.content.firstElementChild;
		}

		if (def != null) { this.select(def, false); }
	}

	/**
	* Select an option.
	*
	* @param {string}  option         The option to select.
	* @param {boolean} dispatch_event Whether to dispatch the
	*                                 component.dropselect.select event on
	*                                 the DropSelect container.
	*
	* @throws {AssertError} If no options are set.
	* @throws {AssertError} If the requested option doesn't exist.
	*/
	select(option, dispatch_event) {
		Assert.assert(this.options != null, 'No options set.');
		Assert.assert(this.options.includes(option), 'No such option.');

		let elems = null;

		this.selected = option;
		elems = [...this.container.querySelector('.dropselect-menu').children];
		for (let e of elems) {
			if (e.getAttribute('value') === option) {
				e.innerHTML = `
					<i class="fas fa-angle-right"
						style="padding-right: 5px;"></i>
					${e.getAttribute('value')}
				`;
			}
		}
		if (dispatch_event) {
			this.container.dispatchEvent('component.dropselect.select');
		}
	}

	/**
	* Get the current selection.
	*
	* @return {string|null} The current selection.
	*/
	get_selection() {
		return this.selected;
	}
}
module.exports = DropSelect;
