// ls-shortcut/shortcut.js
/*
*  A system for easily defining shortucts for the LibreSignage webapp.
*/
var Util = require('libresignage/util/Util');
var assert = require('ls-assert').assert;

module.exports = class Shortcut {
	/*
	*  A class for defining the actual shortcuts.
	*
	*  The only constructor argument 'settings' should be an
	*  object with thr following keys.
	*
	*  keys  = An array of keys for the shortcut. The key names
	*          should match the JavaScript key names.
	*  hook  = The hook function to execute for the shortcut.
	*  defer = A function used for checking whether to execute
	*          the attached function or not. The function is
	*          executed if defer() == true when the shortcut
	*          is pressed.
	*/
	constructor(settings) {
		assert(
			settings === Object(settings),
			'Invalid non-object settings for Shortcut.'
		);
		assert(
			Util.sets_eq(
				Object.keys(settings),
				['keys', 'hook', 'defer']
			),
			'Missing keys in Shortcut settings object.',
		);
		this.keys  = settings.keys;
		this.hook  = settings.hook;
		this.defer = settings.defer;
	}

	matches(keys) { return Util.sets_eq(keys, this.keys); }

	exec() {
		if (!this.defer || !this.defer()) {
			this.hook();
		} else {
			console.log(`Shortcut: '${this.keys.join(' + ')}' deferred.`);
		}
	}
}
